File: docs/identities/get-started/_common/code-examples/go/login_handler.go
package main

import (
	"io"
	"net/http"
)

// LoginHandler handles the /login route
func (app *App) loginHandler(writer http.ResponseWriter, request *http.Request) {
	// Get cookies from the request
	cookies := request.Header.Get("Cookie")

	// Try to verify session with Ory
	session, response, err := app.ory.FrontendAPI.ToSession(request.Context()).Cookie(cookies).Execute()

	// If there's an error or session is not active, redirect to login UI
	if err != nil || (err == nil && !*session.Active) {
		http.Redirect(writer, request, app.tunnelUrl+"/self-service/login/browser", http.StatusSeeOther)
		return
	}

	// If session is valid, send the session data as JSON response
	writer.Header().Set("Content-Type", "application/json")
	writer.WriteHeader(http.StatusOK)
	// Use io.Copy to copy the response body to the writer
	io.Copy(writer, response.Body)
}

File: docs/identities/get-started/_common/code-examples/java/LoginHandler.java
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;
import sh.ory.ApiException;
import sh.ory.api.FrontendApi;
import sh.ory.model.Session;

import java.io.IOException;

@RestController
public class LoginHandler {
    private final FrontendApi ory;
    private final String baseUrl;

    public LoginHandler(FrontendApi ory, String baseUrl) {
        this.ory = ory;
        this.baseUrl = baseUrl;
    }

    @GetMapping("/login")
    public void login(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String cookieHeader = request.getHeader("Cookie");

        try {
            Session session = ory.toSession(null, cookieHeader, null);
            if (session != null && session.getActive() != null && session.getActive()) {
                // Session is valid, return session data as JSON
                response.setContentType("application/json");
                response.setStatus(HttpServletResponse.SC_OK);
                response.getWriter().write(session.toString());
                return;
            }
        } catch (ApiException e) {
            // Session is invalid or doesn't exist
        }

        // Redirect to login page
        response.sendRedirect(baseUrl + "/self-service/login/browser");
    }
}
